lib/mat_SDCZ.ml

```diff
@@ -158,6 +158,63 @@ let to_col_vecs mat =
     for i = 2 to n do ar.(i - 1) <- col mat i done;
     ar
 
+let of_col_vecs_list = function
+  | [] -> empty
+  | (vec :: _) as lst ->
+    let n = List.length lst in
+    let m = Array1.dim vec in
+    let mat = create m n in
+    let rec loop i = function
+      | [] -> mat
+      | vec :: t ->
+        if Array1.dim vec <> m then
+          failwith "of_col_vecs_list: vectors not of same length";
+        direct_copy ~n:m ~ofsy:1 ~incy:1 ~y:(col mat i) ~ofsx:1 ~incx:1 ~x:vec;
+        loop (i + 1) t
+    in
+    loop 1 lst
+
+let to_col_vecs_list mat =
+  let n = dim2 mat in
+  let rec loop i acc =
+    if i < 1 then acc
+    else
+      loop (i - 1) (col mat i :: acc)
+  in
+  loop n []
+
+let of_list = function
+  | [] -> empty
+  | (h :: _) as lst ->
+    let m = List.length lst in
+    let n = List.length h in
+    let mat = create m n in
+    let rec loop i = function
+      | [] -> mat
+      | h :: t ->
+        let s = List.fold_left (fun j e -> mat.{i, j} <- e; j + 1) 1 h in
+        if s <= n then
+          invalid_arg "list not long enough"
```

Comment:
I use `Invalid_argument` in this case, because the indexing operator `{i,j}` also throws `Invalid_argument`.  I explicitly catch this exception when I wrap the the base call to `loop` a couple of lines below. At which point I raise `Failure` with ```failwith "of_list: vectors not of same length" ```. I am being consistent with the other functions.

Regardless, this code was too cute from the beginning. I was thinking that the bottleneck in this function would be the double traversal of the lists. I was wrong, as I tested in this [gist](https://gist.github.com/rleonid/c331caf29e88f78abfa2.js). I'll commit a version which counts first and then `failwith`s appropriately in a little bit.